nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-processors/src/main/java/org/apache/nifi/processors/cassandra/QueryCassandra.java

```diff
@@ -217,60 +243,94 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
+                                    nrOfRows.set(convertToJsonStream(resultSet, maxRowsPerFlowFile,
+                                            out, charset, 0, null));
+                                }
                             }
+                        } catch (final TimeoutException | InterruptedException | ExecutionException e) {
+                            throw new ProcessException(e);
                         }
-
-                    } catch (final TimeoutException | InterruptedException | ExecutionException e) {
-                        throw new ProcessException(e);
                     }
-                }
-            });
+                });
+
+                // set attribute how many rows were selected
+                fileToProcess = session.putAttribute(fileToProcess, RESULT_ROW_COUNT, String.valueOf(nrOfRows.get()));
 
-            // set attribute how many rows were selected
-            fileToProcess = session.putAttribute(fileToProcess, RESULT_ROW_COUNT, String.valueOf(nrOfRows.get()));
+                // set mime.type based on output format
+                fileToProcess = session.putAttribute(fileToProcess, CoreAttributes.MIME_TYPE.key(),
+                        JSON_FORMAT.equals(outputFormat) ? "application/json" : "application/avro-binary");
 
-            // set mime.type based on output format
-            fileToProcess = session.putAttribute(fileToProcess, CoreAttributes.MIME_TYPE.key(),
-                    JSON_FORMAT.equals(outputFormat) ? "application/json" : "application/avro-binary");
+                logger.info("{} contains {} records; transferring to 'success'",
+                        new Object[]{fileToProcess, nrOfRows.get()});
+                session.getProvenanceReporter().modifyContent(fileToProcess, "Retrieved " + nrOfRows.get() + " rows",
+                        stopWatch.getElapsed(TimeUnit.MILLISECONDS));
+                session.transfer(fileToProcess, REL_SUCCESS);
 
-            logger.info("{} contains {} Avro records; transferring to 'success'",
-                    new Object[]{fileToProcess, nrOfRows.get()});
-            session.getProvenanceReporter().modifyContent(fileToProcess, "Retrieved " + nrOfRows.get() + " rows",
-                    stopWatch.getElapsed(TimeUnit.MILLISECONDS));
-            session.transfer(fileToProcess, REL_SUCCESS);
+                if (outputBatchSize > 0) {
+                    flowFileCount++;
+
+                    if (flowFileCount == outputBatchSize) {
+                        session.commitAsync();
+                        flowFileCount = 0;
+                        fileToProcess = session.create();
+                    }
+                }
+                try {
+                    resultSet.fetchMoreResults().get();
+                } catch (Exception e) {
+                    logger.error("ExecutionException : query {} for {} due to {}; routing to failure",
+                            new Object[]{selectQuery, fileToProcess, e});
```

Comment:
Agree, added another try-catch block. It is a bit clunky in the current iteration, because of the repeated code. Let me know if you want me to try to clean those up.